ftdc/custom_format.go

```diff
@@ -0,0 +1,549 @@
+package ftdc
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/binary"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"reflect"
+	"strings"
+
+	"go.viam.com/rdk/logging"
+)
+
+// epsilon is a small value for determining whether a float is 0.0.
+const epsilon = 1e-9
+
+type Schema struct {
+	// A `Datum`s data is a map[string]any. Even if two datum's have maps with the same keys, we do
+	// not assume ranging over the map will yield the same order. Thus we explicitly write down an
+	// order to walk the map of new data values.
+	//
+	// mapOrder is only used for writing out new data points.
+	mapOrder []string
+
+	// fieldOrder is flattened list of strings representing individual metrics. Fields use a
+	// dot-notation to represent structure/nesting. E.g: "leftMotor.PowerPct".
+	fieldOrder []string
+}
+
+// writeSchema writes down names for metrics in the form of a json array. All subsequent calls to
+// `writeDatum` will assume this "header" representation until the next call to `writeSchema`. A
+// full description of the file format follows:
+//
+// To build intuition for the file format, let's start with a simple example input being serialized
+// with a simple format. Let's say we have two data points (two "datum"):
+//
+// Datum1 = {time: 123, motor: {powerPct: 0.2, pos: 5000}, gps: {lat: 40.7128, long: -74.0060}}
+// Datum2 = {time: 124, motor: {powerPct: 0.2, pos: 5001}, gps: {lat: 40.7128, long: -74.0061}}
+//
+// We can serialize those two data points as pure json into a file. One after the other:
+//
+// {"time", 123, "motor": {"powerPct": 0.2, "pos": 5000}, gps: {"lat": 40.7128, "long": -74.0060}}\n
+// {"time", 124, "motor": {"powerPct": 0.2, "pos": 5001}, gps: {"lat": 40.7128, "long": -74.0061}}
+//
+// While json can be a bit wasteful in its own right, there are two properties of the data we want
+// to collect that allow us greatly compress data:
+//   - The metrics we're recording (e.g: `motor.powerPct`, `gps.lat`, ...) are often exactly the same
+//     between the readings.
+//   - Many of the values we're recording (e.g: the `0.2`) do not change between readings.
+//
+// Thus, the two tricks employed are to:
+// - Only write metric names when things change.
+// - Minimize how much space is used to write out a value that hasn't changed.
+//
+// Using a pseudo EBNF notation, an FTDC file is:
+// FTDC = ftdc_doc*
+//
+// ftdc_doc = schema | metric
+//
+// schema =
+//
+//	schema_identifier : 0x01 (a full byte of value 1)
+//	schema : <array of strings serialized as JSON, including a trailing \n(0xa)>
+//
+// metric_reading =
+//
+//	metric_identifier : 0b0 (a single bit of value 0)
+//	diff_bit : bit*
+//	time: int64 <Golang: `time.Now().Unix()`. Nanoseconds since the 1970 epoch.>
+//	values : float32*
```

Comment:
I don't think i realized until I got to `parse` that it was possible to store ints / other numeric values...

How will you be able to detect the original types of the `metric_reading` `values` given the fact that the schema doesn't describe the types of the columns?

Reply:
As mentioned elsewhere, we don't save type information. The format is lossy in a way we don't expect to be problematic.

Probably worth me calling out in this documentation somewhere...